source/extensions/filters/http/lua/lua_filter.cc

```diff
@@ -499,6 +500,26 @@ int StreamHandleWrapper::luaVerifySignature(lua_State* state) {
   return 2;
 }
 
+int StreamHandleWrapper::luaTimestamp(lua_State* state) {
+  auto now = time_source_.systemTime().time_since_epoch();
+
+  auto time_unit = luaL_optstring(state, 2, "");
+  if (strcmp(time_unit, "") == 0 || strcmp(time_unit, "milliseconds_since_epoch") == 0) {
```

Comment:
Can we do enum (enum in C++, table in lua) here? To accomplish that, I can think probably we can add an additional initializer function when we initialize the `lua_state_`. This initializer basically populate global state with e.g. a global table.

Reply:
Hm, were you thinking of initializing it here? https://github.com/mindyor/envoy/blob/master/source/extensions/filters/http/lua/lua_filter.cc#L518

`lua_newtable`/`lua_createtable` want a `lua_State`, but I only see `lua_state_` which is a `ThreadLocalState`. Is there a way to unwrap lua_State from here? I'm all ears!

Reply:
I put a quick hack: https://github.com/envoyproxy/envoy/compare/master...dio:lua-enum?expand=1. Sure you can hack it up to create a `registerEnum()` function too.

By having that,

I managed to log the enums, within each global "hook" function that we have:

```yaml
          - name: envoy.filters.http.lua
            typed_config:
              "@type": type.googleapis.com/envoy.config.filter.http.lua.v2.Lua
              inline_code: |
                function envoy_on_request(request_handle)
                  request_handle:logInfo(tostring(EnvoyTimestampResolution.MILLISECOND))
                  request_handle:logInfo(tostring(EnvoyTimestampResolution.NANOSECOND))
                end
                function envoy_on_response(response_handle)
                end
```

Reply:
This is the way I register it. A bit ugly.

```c++
          [](lua_State* state) {
            lua_newtable(state);
            {
              LUA_ENUM(state, MILLISECOND, Timestamp::Resolution::Millisecond);
              LUA_ENUM(state, NANOSECOND, Timestamp::Resolution::Nanosecond);
            }
            lua_setglobal(state, "EnvoyTimestampResolution");
          },
```

Reply:
Hm, I think I see what you're saying, let me give it a shot.

Reply:
Pushing an implementation per your sketch (thank you for all of the detail!). I will try to come back to this PR a bit later this week (getting pulled into other things) to refactor it.